app/src/main/java/com/imashnake/animite/data/Resource.kt

```diff
@@ -0,0 +1,67 @@
+package com.imashnake.animite.data
+
+import com.apollographql.apollo3.ApolloCall
+import com.apollographql.apollo3.api.ApolloResponse
+import com.apollographql.apollo3.api.Operation
+import com.apollographql.apollo3.exception.ApolloCompositeException
+import com.apollographql.apollo3.exception.ApolloNetworkException
+import com.apollographql.apollo3.exception.CacheMissException
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.map
+
+sealed class Resource<T>(
+    open val data: T?,
+    open val message: String? = null
+) {
+
+    data class Success<T>(override val data: T) : Resource<T>(data)
+    data class Error<T>(override val message: String?, override val data: T? = null) : Resource<T?>(data, message)
+    class Loading<T> : Resource<T>(null)
+
+    companion object {
+        fun <T> success(data: T): Resource<T> {
+            return Success(data)
+        }
+
+        fun <T> error(msg: String, data: T? = null): Resource<T?> {
+            return Error(msg, data)
+        }
+
+        fun <T> loading(): Resource<T> {
+            return Loading()
+        }
+
+        /**
+         * Parses an [ApolloCall] into a [Resource] by checking its data & errors
+         * As well as catching non-HTTP exceptions like no hostname for no connection & cache miss for cache-only requests
+         *
+         * TODO: Should require a mapper rather than a higher order lambda to provide access to the [ApolloResponse.data]
+         */
+        fun <T : Operation.Data, R> Flow<ApolloResponse<T>>.asResource(mapper: (T) -> R = { it as R }): Flow<Resource<R?>> {
```

Comment:
I only way I can remove this cast is to remove the default parameter from this, which I can do for now, just wanted a way for it to be used with `.asResource()` when we use the root graphQL model rather than `MediaListQuery.Page` or `MediaQuery.Media?`

Reply:
So either we get `.asResource()` or we require `.asResource { it }`